usr/lib/common/pkey_utils.c

```diff
@@ -0,0 +1,531 @@
+    register const unsigned char *__src __asm__("2") = src;
+    register long __src_len __asm__("3") = src_len;
+    register unsigned char *__dest __asm__("4") = dest;
+
+    __asm__ volatile (
+        "0: .insn   rre, 0xb92f0000,%2,%0 \n"   /* KMC opcode */
+        "   brc 1, 0b \n"   /* handle partial completion */
+        : "+a"(__src), "+d"(__src_len), "+a"(__dest)
+        : "d"(__func), "a"(__param)
+        : "cc", "memory");
+
+    return func ? src_len - __src_len : __src_len;
+}
+
+/**
+ * s390_kmac:
+ * @func: the function code passed to KMAC; see s390_kmac_func
+ * @param: address of parameter block; see POP for details on each func
+ * @src: address of source memory area
+ * @src_len: length of src operand in bytes
+ *
+ * Executes the KMAC (COMPUTE MESSAGE AUTHENTICATION CODE) operation of the CPU.
+ *
+ * Returns -1 for failure, 0 for the query func, number of processed
+ * bytes for encryption/decryption funcs
+ */
+static int s390_kmac(unsigned long func, void *param,
+            const unsigned char *src, long src_len)
+{
+    register long __func __asm__("0") = func;
+    register void *__param __asm__("1") = param;
+    register const unsigned char *__src __asm__("2") = src;
+    register long __src_len __asm__("3") = src_len;
+
+    __asm__ volatile (
+        "0:     .insn   rre, 0xb91e0000,%0,%0 \n"
+        "       brc     1, 0b \n"
+        : "+a"(__src), "+d"(__src_len)
+        : "d"(__func), "a"(__param)
+        : "cc", "memory");
+    return func ? src_len - __src_len : __src_len;
+}
+
+/**
+ * s390_pcc:
+ * @func: the function code passed to KM; see s390_pcc_functions
+ * @param: address of parameter block; see POP for details on each func
+ *
+ * Executes the PCC operation of the CPU.
+ *
+ * Returns -1 for failure, 0 for the query func, number of processed
+ * bytes for encryption/decryption funcs
+ */
+static int s390_pcc(unsigned long func, void *param)
+{
+    register unsigned long r0 __asm__("0") = (unsigned long)func;
+    register unsigned long r1 __asm__("1") = (unsigned long)param;
+
+    __asm__ volatile (
+        "0: .long   %[opc] << 16\n"
+        "   brc 1,0b\n"
+        :
+        : [fc] "d" (r0), [param] "a" (r1), [opc] "i" (0xb92c)
+        : "cc", "memory");
+
+    return 0;
+}
+
+/**
+ * Update the specified attribute of the given key object. The object gets
+ * locked for write and is saved, if it's a token object.
+ *
+ * Note: When calling this function, the XProcLock MUST NOT be held,
+ *       because it tries to obtain a write lock on the key object.
+ */
+CK_RV pkey_update_and_save(STDLL_TokData_t *tokdata, OBJECT *key_obj,
+                           CK_ATTRIBUTE *pkey_attr)
+{
+    CK_RV ret;
+
+    /* Unlock obj to remove READ_LOCK */
+    ret = object_unlock(key_obj);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("object_unlock failed with rc=0x%lx\n", ret);
+        return ret;
+    }
+
+    /* Obtain write lock on key object */
```

Comment:
Since you have unlocked the object here, it could be modified by other threads until you get the write lock. 

Could there be any problem when other threads modify the object or its attributes in any way here? 
This function will blindly add/update the PKEY attribute, so this function should be OK. 
But what about the caller(s) of this function? Do they somehow keep a reference on an attribute or attribute value of this object? This reference may no longer be valid on return of this function (due to unlock of the object and concurrent modification).

What happens if multiple threads try to set a new PKEY for the same object? I guess the last one will win, and the prot key will anyway be all the same, but please check carefully if such multithreading issues are cleanly handled..

Reply:
Routine ep11tok_pkey_usage_ok is now reworked and renamed to ep11tok_pkey_check. (The second routine, which was ep11tok_pkey_usage_ok2, is now renamed to ep11tok_pkey_usage_ok).